docs/design/2020-09-30-index-usage-information.md

```diff
@@ -0,0 +1,159 @@
+# Proposal:
+
+- Author(s):     [rebelice](https://github.com/rebelice)
+- Last updated:  Sep. 30, 2020
+- Discussion at: N/A
+
+## Abstract
+
+This document describes the design of recording index usage information.
+
+## Background
+
+There may be unused indexes in the database. In addition, modifying database tables, indexes, and query statements may cause some indexes to not be used in the future. Users may want to clear these long-unused indexes to relieve storage and performance pressure.
+
+Related issues:
+
+- https://github.com/pingcap/tidb/issues/14998
+- https://github.com/pingcap/tidb/issues/14607
+- https://github.com/pingcap/tidb/issues/17508
+- https://asktug.com/t/topic/34614/12
+
+## Proposal
+
+### SCHEMA_INDEX_USAGE
+
+Design system tables to record index usage information. The system table is designed as follows:
+
+1. The system table is located in `mysql` database and is named `SCHEMA_INDEX_USAGE`.
+2. Columns of `SCHEMA_INDEX_USAGE`:
+
+| Column name  | Data type | Description |
+|--------------|-----------|-------------|
+| TABLE_SCHEMA | varchar   | Name of the database on which the table or view is defined.|
+| TABLE_NAME   | varchar   | Name of the table or view on which the index is defined.|
+| INDEX_NAME   | varchar   | Name of the index.|
+| QUERY_COUNT  | longlong  | Number of the SQL using this index.|
+| ROWS_SELECTED| longlong  | Number of rows read from the index. We can check the selectivity of the index through `ROWS_READ` / `QUERY_COUNT`.|
+| LAST_USED_AT | timestamp | The last time of the SQL using this index.|
+
+
+3. Typical usage: `select * from mysql.SCHEMA_INDEX_USAGE`.
+
+#### Table creation: 
+
+```sql
+create table SCHEMA_INDEX_USAGE (
+	TABLE_SCHEMA varchar(64),
+	TABLE_NAME varchar(64),
+	INDEX_NAME varchar(64),
+	QUERY_COUNT BIGINT,
+	ROUWS_SELECTED BIGINT,
+	LAST_USED_AT timestamp,
+	PRIMARY KEY (SCHEMA_NAME, TABLE_NAME, INDEX_NAME)
+);
+```
+
+Because the max length of schema name, table name and index name is 64, so we use `varchar(64)`. The documentation is [here](https://dev.mysql.com/doc/refman/5.7/en/identifier-length.html).
+
+#### Table update: 
+
+```sql
+insert into mysql.SCHEMA_INDEX_USAGE values (schemaA, tableA, indexA, count, row, used_at) 
+on duplicate key update query_count=query_count+count, rows_selected=rows_selected+row, last_used_at=greatest(last_used_at, used_at)
+```
+
+#### Update method:
+
+Statistics update is divided into statistics and persistence of index usage information. Index usage information is counted by the exec-info collector. And persistence is periodically writing data to system tables. We add a global variable `index-usage-sync-lease` to control the persistence cycle. It is set to 1 minute by default. In addition, add a global SQL variable to control whether to turn on this feature.
+
+### SCHEMA_UNUSED_INDEXES
+
+Due to MySQL compatibility, add the system table `SCHEMA_UNUSED_INDEXES`.
```

Comment:
MySQL's `SCHEMA_UNUSED_INDEXES` is under `SYS` database, should we also be compatible with that?

Reply:
Sounds reasonable. But I think create a `SYS` database is a big issue and may need to be discussed as a new, separate matter. What do you think?